component/loki/source/aws_firehose/component.go

```diff
@@ -0,0 +1,173 @@
+
+	opts component.Options
+	args Arguments
+
+	// utils
+	serverMetrics *util.UncheckedCollector
+	logger        log.Logger
+}
+
+// New creates a new Component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	c := &Component{
+		opts:          o,
+		destination:   make(loki.LogsReceiver),
+		fanout:        args.ForwardTo,
+		serverMetrics: util.NewUncheckedCollector(nil),
+
+		logger: log.With(o.Logger, "component", "aws_firehose_logs"),
+	}
+
+	o.Registerer.MustRegister(c.serverMetrics)
+
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run starts a routine forwards received message to each destination component.
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+		c.shutdownServer()
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.destination:
+			c.mut.RLock()
+			for _, receiver := range c.fanout {
+				receiver <- entry
+			}
+			c.mut.RUnlock()
+		}
+	}
+}
+
+// Update updates the component with a new configuration, restarting the server if needed.
+func (c *Component) Update(args component.Arguments) error {
+	var err error
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	newArgs := args.(Arguments)
+	c.fanout = newArgs.ForwardTo
+
+	var newRelabels []*relabel.Config = nil
+	var relabelRulesChanged = false
+
+	// todo(pablo): is it a good practice to keep a reference to the arguments in the
+	// component struct, used for comparing here rather than destructuring them?
+	if newArgs.RelabelRules != nil && len(newArgs.RelabelRules) > 0 {
+		relabelRulesChanged = true
+		newRelabels = flow_relabel.ComponentToPromRelabelConfigs(newArgs.RelabelRules)
+	} else if c.rbs != nil && len(c.rbs) > 0 && (newArgs.RelabelRules == nil || len(newArgs.RelabelRules) == 0) {
+		// nil out relabel rules if they need to be cleared
+		relabelRulesChanged = true
+	}
+
+	// Since the handler is created ad-hoc for the server, and the handler depends on the relabels
+	// consider this as a cause for server restart as well. Much simpler than adding a lock on the
+	// handler and doing the relabel rules change on the fly
+	serverNeedsUpdate := !reflect.DeepEqual(c.args.Server, newArgs.Server)
+	if !serverNeedsUpdate && !relabelRulesChanged {
+		c.args = newArgs
+		return nil
+	}
+
+	c.shutdownServer()
+
+	// update relabel rules in component if needed
+	if relabelRulesChanged {
+		c.rbs = newRelabels
+	}
+
+	jobName := strings.Replace(c.opts.ID, ".", "_", -1)
+
+	registry := prometheus.NewRegistry()
+	c.serverMetrics.SetCollector(registry)
+
+	c.server, err = fnet.NewTargetServer(c.logger, jobName, registry, newArgs.Server)
+	if err != nil {
+		return err
+	}
+
+	if err = c.server.MountAndRun(func(router *mux.Router) {
+		// re-create handler when server is re-computed
+		handler := internal.NewHandler(c, c.logger, registry, c.rbs)
+		router.Path("/awsfirehose/api/v1/push").Methods("POST").Handler(handler)
```

Comment:
IMO we should make this pather configurable, with this has the default. I dont see where we are prepending the component id. If we dont make it configurable then it needs to be a singleton and that we are trying to avoid singletons.

Reply:
Regarding the path, just following what we've been using in promtail (hence moving into flow as well). If you check the rest of the `loki.source` components that expose a network server:
- [api](https://grafana.com/docs/agent/next/flow/reference/components/loki.source.api/): `/api/v1/push`
- [heroku](https://grafana.com/docs/agent/latest/flow/reference/components/loki.source.heroku/): `/heroku/api/v1/drain`
- [gcplogs](https://grafana.com/docs/agent/next/flow/reference/components/loki.source.gcplog/#push-block): `/gcp/api/v1/push`

Following that pattern I chose `/awsfirehose/api/v1/push`

What do you mean that it needs to be a singleton?

Reply:
@mattdurham AFAICT It's not a singleton because each instance of the component has its own HTTP server

Reply:
Its a bit painful to setup multiple ones (mainly due to weaveworks always creating a grpc server) but it works. So resolving this.